chainstate/tx-verifier/src/transaction_verifier/hierarchy.rs

```diff
@@ -98,7 +98,7 @@ impl<'a, S: TransactionVerifierStorageRef> UtxosStorageRead for TransactionVerif
         &self,
         id: Id<Block>,
     ) -> Result<Option<utxo::BlockUndo>, storage_result::Error> {
-        match self.utxo_block_undo.get(&id) {
+        match self.utxo_block_undo.get(&TransactionSource::Chain(id)) {
```

Comment:
So the problem is the same as with panics: interfaces `TransactionVerifierStorageRef` and `TransactionVerifierStorageMut` connect both `TransactionVerifier` -> `TransactionVerifier` and `TransactionVerifier` -> `ChainstateRef`.  I believe we want to allow flushing/reading undo data from one verifier to another and forbid it in case of ChainstateRef.